src/libopensc/card-jacartapki.c

```diff
@@ -0,0 +1,1905 @@
+
+	chv_ref = 0;
+	entry = sc_file_get_acl_entry(pin_file, SC_AC_OP_PIN_CHANGE);
+	if (entry != NULL) {
+#if defined(ENABLE_SM)
+		rv = jacartapki_sm_chv_change(card, data, chv_ref, tries_left, entry->key_ref);
+		LOG_FUNC_RETURN(ctx, rv);
+#else
+		LOG_ERROR_RET(ctx, SC_ERROR_NOT_SUPPORTED, "UPDATE CHV/SCB present. PIN change is not supported w/o SM");
+#endif
+	} else {
+		LOG_ERROR_RET(ctx, SC_ERROR_NOT_SUPPORTED, "PIN change is not supported for KO object w/o UPDATE CHV/SCB");
+	}
+}
+
+static int
+jacartapki_pin_reset(struct sc_card *card, struct sc_pin_cmd_data *data, int *tries_left)
+{
+	struct sc_context *ctx = card->ctx;
+	struct sc_apdu apdu;
+	struct sc_file *pin_file = NULL;
+	int rv;
+
+	LOG_FUNC_CALLED(ctx);
+	sc_log(ctx, "Reset PIN(type:%i,ref:%i,lengths:%" SC_FORMAT_LEN_SIZE_T "u/%" SC_FORMAT_LEN_SIZE_T "u)", data->pin_type, data->pin_reference, data->pin1.len, data->pin2.len);
+
+	if (data->pin_type != SC_AC_CHV)
+		LOG_FUNC_RETURN(ctx, SC_ERROR_NOT_SUPPORTED);
+
+	if (!(data->pin_reference & 0x80)) {
+		const struct sc_acl_entry *entry;
+
+		rv = jacartapki_select_global_pin(card, data->pin_reference, &pin_file);
+		LOG_TEST_RET(ctx, rv, "Select PIN file error");
+
+		if (data->pin1.len) {
+			entry = sc_file_get_acl_entry(pin_file, SC_AC_OP_PIN_RESET);
+			if (entry != NULL) {
+				sc_log(ctx, "Acl(PIN_RESET): %04X", entry->key_ref);
+				if ((entry->key_ref & 0x00FF) == 0xFF) {
+					LOG_ERROR_RET(ctx, SC_ERROR_NOT_ALLOWED, "Reset PIN not allowed");
+				} else if ((entry->key_ref & 0xC000) != 0) {
+					LOG_ERROR_RET(ctx, SC_ERROR_NOT_SUPPORTED, "Reset PIN protected by SM: not supported (TODO)");
+				} else if ((entry->key_ref & 0x00FF) != 0) {
+					rv = jacartapki_pin_verify(card, SC_AC_CHV, entry->key_ref & 0x00FF, data->pin1.data, data->pin1.len, tries_left);
+					LOG_TEST_RET(ctx, rv, "Verify PUK failed");
+
+					rv = jacartapki_select_global_pin(card, data->pin_reference, &pin_file);
+					LOG_TEST_RET(ctx, rv, "Select PIN file error");
+				}
+			}
+		}
+	}
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_1, 0x2C, 0, 0);
+	apdu.cla = 0x80;
+
+	rv = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(ctx, rv, "APDU transmit failed");
+	rv = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	LOG_TEST_RET(ctx, rv, "PIN change failed");
+
+	if (data->pin2.len > 0) {
+		size_t save_len = data->pin1.len;
+
+		data->pin1.len = 0;
+		rv = jacartapki_pin_change(card, data, tries_left);
+		data->pin1.len = save_len;
+		LOG_TEST_RET(ctx, rv, "Cannot set new PIN value");
+	}
+
+	LOG_FUNC_RETURN(ctx, SC_SUCCESS);
+}
+
+static int
+jacartapki_pin_getinfo(struct sc_card *card, struct sc_pin_cmd_data *data)
+{
+	size_t i;
+	struct jacartapki_private_data *private_data = (struct jacartapki_private_data *)card->drv_data;
+	for (i = 0; i < ARRAY_SIZE(private_data->auth_state); ++i) {
+		if (private_data->auth_state[i].pin_reference == (unsigned int)data->pin_reference) {
+			data->pin1.logged_in = private_data->auth_state[i].logged_in;
```

Comment:
when assigning `data->pin1.logged_in`, you need to use the definitions from `opensc.h`. When setting `private_data->auth_state[i].logged_in`, you are currently using the magic values `0` or `1`. A possible problem may arise from #3490